File: src/org/stringtemplate/v4/compiler/Compiler.java
/*
 * [The "BSD license"]
 *  Copyright (c) 2011 Terence Parr
 *  All rights reserved.
 *
 *  Redistribution and use in source and binary forms, with or without
 *  modification, are permitted provided that the following conditions
 *  are met:
 *  1. Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *  2. Redistributions in binary form must reproduce the above copyright
 *     notice, this list of conditions and the following disclaimer in the
 *     documentation and/or other materials provided with the distribution.
 *  3. The name of the author may not be used to endorse or promote products
 *     derived from this software without specific prior written permission.
 *
 *  THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 *  IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 *  OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 *  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 *  NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 *  DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 *  THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 *  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 *  THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
package org.stringtemplate.v4.compiler;

import java.util.Collections;
import java.util.concurrent.atomic.AtomicInteger;
import org.antlr.runtime.ANTLRStringStream;
import org.antlr.runtime.CommonTokenStream;
import org.antlr.runtime.NoViableAltException;
import org.antlr.runtime.Parser;
import org.antlr.runtime.RecognitionException;
import org.antlr.runtime.Token;
import org.antlr.runtime.TokenStream;
import org.antlr.runtime.tree.CommonTreeNodeStream;
import org.stringtemplate.v4.Interpreter;
import org.stringtemplate.v4.ST;
import org.stringtemplate.v4.STGroup;
import org.stringtemplate.v4.misc.ErrorType;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

/** A compiler for a single template. */
public class Compiler {
    public static final String SUBTEMPLATE_PREFIX = "_sub";

    public static final int TEMPLATE_INITIAL_CODE_SIZE = 15;

    public static final Map<String, Interpreter.Option> supportedOptions;

    static {
        final Map<String, Interpreter.Option> map = new HashMap<String, Interpreter.Option>();
        map.put("anchor",    Interpreter.Option.ANCHOR);
        map.put("format",    Interpreter.Option.FORMAT);
        map.put("null",      Interpreter.Option.NULL);
        map.put("separator", Interpreter.Option.SEPARATOR);
        map.put("wrap",      Interpreter.Option.WRAP);
        supportedOptions = Collections.unmodifiableMap(map);
    }

    public static final int NUM_OPTIONS = supportedOptions.size();

    public static final Map<String,String> defaultOptionValues;

    static {
        final Map<String, String> map = new HashMap<String, String>();
        map.put("anchor", "true");
        map.put("wrap",   "\n");
        defaultOptionValues = Collections.unmodifiableMap(map);
    }

    public static Map<String, Short> funcs;

    static {
        final Map<String, Short> map = new HashMap<String, Short>();
        map.put("first", Bytecode.INSTR_FIRST);
        map.put("last", Bytecode.INSTR_LAST);
        map.put("rest", Bytecode.INSTR_REST);
        map.put("trunc", Bytecode.INSTR_TRUNC);
        map.put("strip", Bytecode.INSTR_STRIP);
        map.put("trim", Bytecode.INSTR_TRIM);
        map.put("length", Bytecode.INSTR_LENGTH);
        map.put("strlen", Bytecode.INSTR_STRLEN);
        map.put("reverse", Bytecode.INSTR_REVERSE);
        funcs = Collections.unmodifiableMap(map);
    }

    /** Name subtemplates {@code _sub1}, {@code _sub2}, ... */
    public static AtomicInteger subtemplateCount = new AtomicInteger(0);

    public STGroup group;

    public Compiler() { this(STGroup.defaultGroup); }
    public Compiler(STGroup group) { this.group = group; }

    public CompiledST compile(String template) {
        CompiledST code = compile(null, null, null, template, null);
        code.setHasFormalArgs(false);
        return code;
    }

    /** Compile full template with unknown formal arguments. */
    public CompiledST compile(String name, String template) {
        CompiledST code = compile(null, name, null, template, null);
        code.setHasFormalArgs(false);
        return code;
    }

    /** Compile full template with respect to a list of formal arguments. */
    public CompiledST compile(String srcName,
                              String name,
                              List<FormalArgument> args,
                              String template,
                              Token templateToken)
    {
        ANTLRStringStream is = new ANTLRStringStream(template);
        is.name = srcName!=null ? srcName : name;
        STLexer lexer;
        if ( templateToken!=null &&
             templateToken.getType() == GroupParser.BIGSTRING_NO_NL )
        {
            lexer = new STLexer(group.errMgr, is, templateToken,
                        group.delimiterStartChar, group.delimiterStopChar) {
                /** Throw out \n and indentation tokens inside BIGSTRING_NO_NL */
                @Override
                public Token nextToken() {
                    Token t = super.nextToken();
                    while ( t.getType() == STLexer.NEWLINE ||
                            t.getType()==STLexer.INDENT )
                    {
                        t = super.nextToken();
                    }
                    return t;
                }
            };
        }
        else {
            lexer = new STLexer(group.errMgr, is, templateToken,
                                group.delimiterStartChar, group.delimiterStopChar);
        }
        CommonTokenStream tokens = new CommonTokenStream(lexer);
        STParser p = new STParser(tokens, group.errMgr, templateToken);
        STParser.templateAndEOF_return r;
        try {
            r = p.templateAndEOF();
        }
        catch (RecognitionException re) {
            reportMessageAndThrowSTException(tokens, templateToken, p, re);
            return null;
        }
        if ( p.getNumberOfSyntaxErrors()>0 || r.getTree()==null ) {
            CompiledST impl = new CompiledST();
            impl.defineFormalArgs(args);
            return impl;
        }

        //System.out.println(((CommonTree)r.getTree()).toStringTree());
        CommonTreeNodeStream nodes = new CommonTreeNodeStream(r.getTree());
        nodes.setTokenStream(tokens);
        CodeGenerator gen = new CodeGenerator(nodes, group.errMgr, name, template, templateToken);

        CompiledST impl=null;
        try {
            impl = gen.template(name,args);
            impl.setNativeGroup(group);
            impl.setTemplate(template);
            impl.setAST(r.getTree());
            impl.getAST().setUnknownTokenBoundaries();
            impl.setTokens(tokens);
        }
        catch (RecognitionException re) {
            group.errMgr.internalError(null, "bad tree structure", re);
        }

        return impl;
    }

    public static CompiledST defineBlankRegion(CompiledST outermostImpl, Token nameToken) {
        String outermostTemplateName = outermostImpl.getName();
        String mangled = STGroup.getMangledRegionName(outermostTemplateName, nameToken.getText());
        CompiledST blank = new CompiledST();
        blank.setRegion(true);
        blank.setTemplateDefStartToken(nameToken);
        blank.setRegionDefType(ST.RegionType.IMPLICIT);
        blank.setName(mangled);
        outermostImpl.addImplicitlyDefinedTemplate(blank);
        return blank;
    }

    public static String getNewSubtemplateName() {
        int count = subtemplateCount.incrementAndGet();
        return SUBTEMPLATE_PREFIX+count;
    }

    protected void reportMessageAndThrowSTException(TokenStream tokens, Token templateToken,
                                                    Parser parser, RecognitionException re)
    {
        if ( re.token.getType() == STLexer.EOF_TYPE ) {
            String msg = "premature EOF";
            group.errMgr.compileTimeError(ErrorType.SYNTAX_ERROR, templateToken, re.token, msg);
        }
        else if ( re instanceof NoViableAltException) {
            String msg = "'"+re.token.getText()+"' came as a complete surprise to me";
            group.errMgr.compileTimeError(ErrorType.SYNTAX_ERROR, templateToken, re.token, msg);
        }
        else if ( tokens.index() == 0 ) { // couldn't parse anything
            String msg = "this doesn't look like a template: \""+tokens+"\"";
            group.errMgr.compileTimeError(ErrorType.SYNTAX_ERROR, templateToken, re.token, msg);
        }
        else if ( tokens.LA(1) == STLexer.LDELIM ) { // couldn't parse expr
            String msg = "doesn't look like an expression";
            group.errMgr.compileTimeError(ErrorType.SYNTAX_ERROR, templateToken, re.token, msg);
        }
        else {
            String msg = parser.getErrorMessage(re, parser.getTokenNames());
            group.errMgr.compileTimeError(ErrorType.SYNTAX_ERROR, templateToken, re.token, msg);
        }
        throw new STException(); // we have reported the error, so just blast out
    }

}

File: objc/main/compiler/Compiler.m
/*
 * [The "BSD license"]
 *  Copyright (c) 2011 Terence Parr and Alan Condit
 *  All rights reserved.
 *
 *  Redistribution and use in source and binary forms, with or without
 *  modification, are permitted provided that the following conditions
 *  are met:
 *  1. Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *  2. Redistributions in binary form must reproduce the above copyright
 *     notice, this list of conditions and the following disclaimer in the
 *     documentation and/or other materials provided with the distribution.
 *  3. The name of the author may not be used to endorse or promote products
 *     derived from this software without specific prior written permission.
 *
 *  THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 *  IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 *  OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 *  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 *  NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 *  DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 *  THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 *  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 *  THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
#import <Foundation/Foundation.h>
#import <ANTLR/ANTLR.h>
#import "STErrorListener.h"
#import "Compiler.h"
#import "CodeGenerator.h"
#import "ST.h"
#import "CompiledST.h"
#import "Bytecode.h"
#import "Interpreter.h"
#import "STException.h"
#import "ErrorManager.h"
#import "STLexer.h"
#import "STParser.h"
#import "GroupParser.h"

@implementation Compiler_Anon1

+ (id) newCompiler_Anon1
{
    return [[Compiler_Anon1 alloc] init];
}

- (id) init
{
    if ( (self=[super init]) != nil ) {
        dict = [[LinkedHashMap newLinkedHashMap:16] retain];
        [dict put:@"anchor"    value:[ACNumber numberWithInteger:ANCHOR]];
        [dict put:@"format"    value:[ACNumber numberWithInteger:FORMAT]];
        [dict put:@"null"      value:[ACNumber numberWithInteger:_NULL]];
        [dict put:@"separator" value:[ACNumber numberWithInteger:SEPARATOR]];
        [dict put:@"wrap"      value:[ACNumber numberWithInteger:WRAP]];
    }
    return self;
}

- (void) dealloc
{
#ifdef DEBUG_DEALLOC
    NSLog( @"called dealloc in Compiler_Anon1" );
#endif
    if ( dict ) [dict release];
    [super dealloc];
}

- (id) getDict
{
    return dict;
}

- (id) get:(id)aKey
{
    return [dict get:aKey];
}

- (void) put:(id)aKey value:(id)anObject
{
    [dict put:aKey value:anObject];
}

- (NSInteger) count
{
    return [dict count];
}

@synthesize dict;
@end

@implementation Compiler_Anon2

+ (id) newCompiler_Anon2
{
    return [[Compiler_Anon2 alloc] init];
}

- (id) init
{
    if ( (self=[super init]) != nil ) {
        dict = [[LinkedHashMap newLinkedHashMap:16] retain];
        [dict put:@"anchor" value:@"\"true\""];
        [dict put:@"wrap" value:@"\"\n\""];
    }
    return self;
}

- (void) dealloc
{
#ifdef DEBUG_DEALLOC
    NSLog( @"called dealloc in Compiler_Anon2" );
#endif
    if ( dict ) [dict release];
    [super dealloc];
}

- (id) copyWithZone:(NSZone *)aZone
{
    Compiler_Anon2 *copy;
    
    copy = [[[self class] allocWithZone:aZone] init];
    if ( dict ) {
        if ( copy.dict ) [copy.dict release];
        copy.dict = [dict copyWithZone:aZone];
        [copy.dict retain];
    }
    return copy;
}

- (id) getDict
{
    return dict;
}

- (id) get:(id)aKey
{
    return [dict get:aKey];
}

- (void) put:(id)aKey value:(id)anObject
{
    [dict put:aKey value:anObject];
}

- (NSInteger) count
{
    return [dict count];
}

@synthesize dict;
@end

@implementation Compiler_Anon3

+ (id) newCompiler_Anon3
{
    return [[Compiler_Anon3 alloc] init];
}

- (id) init
{
    if ( (self=[super init]) != nil ) {
        dict = [[LinkedHashMap newLinkedHashMap:16] retain];
        [dict put:@"first"   value:[NSString stringWithFormat:@"%d", Bytecode.INSTR_FIRST]];
        [dict put:@"last"    value:[NSString stringWithFormat:@"%d", Bytecode.INSTR_LAST]];
        [dict put:@"rest"    value:[NSString stringWithFormat:@"%d", Bytecode.INSTR_REST]];
        [dict put:@"trunc"   value:[NSString stringWithFormat:@"%d", Bytecode.INSTR_TRUNC]];
        [dict put:@"strip"   value:[NSString stringWithFormat:@"%d", Bytecode.INSTR_STRIP]];
        [dict put:@"trim"    value:[NSString stringWithFormat:@"%d", Bytecode.INSTR_TRIM]];
        [dict put:@"length"  value:[NSString stringWithFormat:@"%d", Bytecode.INSTR_LENGTH]];
        [dict put:@"strlen"  value:[NSString stringWithFormat:@"%d", Bytecode.INSTR_STRLEN]];
        [dict put:@"reverse" value:[NSString stringWithFormat:@"%d", Bytecode.INSTR_REVERSE]];
    }
    return self;
}

- (void) dealloc
{
#ifdef DEBUG_DEALLOC
    NSLog( @"called dealloc in Compiler_Anon3" );
#endif
    if ( dict ) [dict release];
    [super dealloc];
}

- (id) copyWithZone:(NSZone *)aZone
{
    Compiler_Anon3 *copy;
    
    copy = [[[self class] allocWithZone:aZone] init];
    if ( dict ) {
        if ( copy.dict ) [copy.dict release];
        copy.dict = [dict copyWithZone:aZone];
        if ( copy.dict ) [copy.dict retain];
    }
    return copy;
}

- (id) getDict
{
    return dict;
}

- (short) getInstr:(NSString *)aKey
{
    return (short)[(NSString *)[dict get:aKey] intValue];
}

- (void) put:(id)aKey setInstr:(short)anInstr
{
    [dict put:aKey value:[NSString stringWithFormat:@"%d", anInstr]];
}

- (NSInteger) count
{
    return [dict count];
}

@synthesize dict;
@end

@implementation Compiler

static Compiler_Anon1 *supportedOptions;
static NSInteger NUM_OPTIONS;
static Compiler_Anon2 *defaultOptionValues;
static Compiler_Anon3 *funcs;
/**
 * Name subtemplates _sub1, _sub2, ...
 */
static NSInteger subtemplateCount = 0;
static NSString *SUBTEMPLATE_PREFIX = @"_sub";
//static NSInteger TEMPLATE_INITIAL_CODE_SIZE = 15;

@synthesize group;

+ (void) initialize
{
    supportedOptions = [Compiler_Anon1 newCompiler_Anon1];
    NUM_OPTIONS = [supportedOptions count];
    defaultOptionValues = [Compiler_Anon2 newCompiler_Anon2];
    funcs = [Compiler_Anon3 newCompiler_Anon3];
}

+ (NSString *) SUBTEMPLATE_PREFIX
{
    return SUBTEMPLATE_PREFIX;
}

+ (NSInteger) supportedOptions
{
    return [supportedOptions count];
}

+ (Compiler_Anon1 *) getSupportedOptions
{
    return supportedOptions;
}

+ (NSInteger) NUM_OPTIONS
{
    return NUM_OPTIONS;
}

+ (Compiler_Anon2 *) defaultOptionValues
{
    return defaultOptionValues;
}

+ (Compiler_Anon3 *) funcs
{
    return funcs;
}

+ (NSInteger) subtemplateCount
{
    return subtemplateCount;
}

+ (Compiler *) newCompiler
{
    return [[Compiler alloc] initWithSTGroup:STGroup.defaultGroup];
}

+ (Compiler *) newCompiler:(STGroup *)aSTGroup
{
    return [[Compiler alloc] initWithSTGroup:aSTGroup];
}

- (id) init
{
    self=[super init];
    if ( self != nil ) {
        group = STGroup.defaultGroup;
        if ( group ) [group retain];
        subtemplateCount = 0;
    }
    return self;
}

- (id) initWithSTGroup:(STGroup *)aSTGroup
{
    self=[super init];
    if ( self != nil ) {
        group = aSTGroup;
        if ( group ) [group retain];
        subtemplateCount = 0;
    }
    return self;
}

- (void) dealloc
{
#ifdef DEBUG_DEALLOC
    NSLog( @"called dealloc in Compiler" );
#endif
    if ( group ) [group release];
    [super dealloc];
}

/**
 * To compile a template, we need to know what the
 * enclosing template is (if any) in case of regions.
 */
- (CompiledST *) compile:(NSString *)template
{
    CompiledST *code = [self compile:nil name:nil args:nil template:template templateToken:nil];
    code.hasFormalArgs = NO;
    return code;
}

/**
 * Compile full template with unknown formal args.
 */
- (CompiledST *) compile:(NSString *)name template:(NSString *)template
{
    CompiledST *code = [self compile:nil name:name args:nil template:template templateToken:nil];
    code.hasFormalArgs = NO;
    return code;
}

/** Compile full template with respect to a list of formal args. */
- (CompiledST *) compile:(NSString *)srcName
                    name:(NSString *)name
                args:(AMutableArray *)args
                template:(NSString *)template
           templateToken:(CommonToken *)aTemplateToken
{
    CompiledST *impl = nil;
    BOOL mustRelease = NO;
    __strong FormalArgument *a;
    if ( args != nil ) {
            [args retain];
        if ( [args count] > 0 ) {
            a = [args objectAtIndex:0];
            [a retain];
            mustRelease = YES;
        }
    }
    if ( args == nil ) {
        NSLog( @"args is nil" );
    }
    ANTLRStringStream *is = [[ANTLRStringStream newANTLRStringStream:template] retain];
    is.name = (srcName != nil) ? srcName : name;
    STLexer *lexer = nil;
	if ( aTemplateToken != nil &&
		 aTemplateToken.type == GroupParser.TBIGSTRING_NO_NL ) {
            lexer = [STLexer_NO_NL newSTLexer_NO_NL:group.errMgr input:is templateToken:aTemplateToken delimiterStartChar:group.delimiterStartChar delimiterStopChar:group.delimiterStopChar];
		}
	else {
        lexer = [STLexer newSTLexer:group.errMgr input:is templateToken:aTemplateToken delimiterStartChar:group.delimiterStartChar delimiterStopChar:group.delimiterStopChar];
	}

    CommonTokenStream *tokens = [[CommonTokenStream newCommonTokenStreamWithTokenSource:lexer] retain];
    STParser *p = [STParser newSTParser:tokens error:group.errMgr token:aTemplateToken];
    STParser_templateAndEOF_return *r = nil;
    
    @try {
        r = [p templateAndEOF];
    }
    @catch (RecognitionException *re) {
        [self reportMessageAndThrowSTException:tokens templateToken:aTemplateToken aParser:p re:re];
        return nil;
    }
    if ([p getNumberOfSyntaxErrors] > 0 || r == nil || r.tree == nil) {
        impl = [CompiledST newCompiledST];
        [impl defineFormalArgs:args];
        return impl;
    }
    CommonTreeNodeStream *nodes = [[CommonTreeNodeStream newCommonTreeNodeStream:r.tree] retain];
    [nodes setTokenStream:tokens];
    CodeGenerator *gen = [CodeGenerator newCodeGenerator:nodes errMgr:group.errMgr name:name template:template token:aTemplateToken];

    @try {
        impl = [[gen template:name arg1:args] retain];
		impl.nativeGroup = group;
        if ( impl.nativeGroup ) [impl.nativeGroup retain];
		impl.template = template;
        if ( template ) [template retain];
        impl.ast = [(CommonTree *)r.tree retain];
        [impl.ast setUnknownTokenBoundaries];
        impl.tokens = [tokens retain];
    }
    @catch (RecognitionException *re) {
        [group.errMgr internalError:nil msg:@"bad tree structure" e:re];
    }
    if ( mustRelease ) [a release];
    return impl;
}

+ (CompiledST *) defineBlankRegion:(CompiledST *)outermostImpl token:(CommonToken *)nameToken
{
    NSString *outermostTemplateName = outermostImpl.name;
    NSString *mangled = [STGroup getMangledRegionName:outermostTemplateName name:nameToken.text];
    CompiledST *blank = [CompiledST newCompiledST];
    blank.isRegion = YES;
    blank.templateDefStartToken = nameToken;
    blank.regionDefType = IMPLICIT;
    blank.name = mangled;
    [outermostImpl addImplicitlyDefinedTemplate:blank];
    return blank;
}

+ (NSString *) getNewSubtemplateName
{
    return [NSString stringWithFormat:@"%@%d", SUBTEMPLATE_PREFIX, ++subtemplateCount];
}

- (void) reportMessageAndThrowSTException:(CommonTokenStream *)tokens
                            templateToken:(CommonToken *)templateToken
                                  aParser:(Parser *)aParser
                                       re:(RecognitionException *)re
{
    NSString *msg;
    if ( re.token.type == STLexer.EOF_TYPE) {
        msg = @"premature EOF";
        [group.errMgr compileTimeError:SYNTAX_ERROR templateToken:templateToken t:re.token arg:msg];
    }
    else if ([re isKindOfClass:[NoViableAltException class]]) {
        msg = [NSString stringWithFormat:@"'%@' came as a complete surprise to me", re.token.text];
        [group.errMgr compileTimeError:SYNTAX_ERROR templateToken:templateToken t:re.token arg:msg];
    }
    else if (tokens.index == 0) {
        msg = [NSString stringWithFormat:@"this doesn't look like a template: \"%@\"", [tokens description]];
        [group.errMgr compileTimeError:SYNTAX_ERROR templateToken:templateToken t:re.token arg:msg];
    }
    else if ([tokens LA:1] == STLexer.LDELIM) {
        msg = @"doesn't look like an expression";
        [group.errMgr compileTimeError:SYNTAX_ERROR templateToken:templateToken t:re.token arg:msg];
    }
    else {
        msg = [aParser getErrorMessage:re TokenNames:[aParser getTokenNames]];
        [group.errMgr compileTimeError:SYNTAX_ERROR templateToken:templateToken t:re.token arg:msg];
    }
    @throw [STException newException:msg];
}

@end
